pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1419,40 +1428,42 @@ private ByteBuf processMessageChunk(ByteBuf compressedPayload, MessageMetadata m
      * Notify waiting asyncReceive request with the received message.
      *
      * @param message
+     * @return return true if this message is handled by pending receive.
      */
-    void notifyPendingReceivedCallback(final Message<T> message, Exception exception) {
-        if (pendingReceives.isEmpty()) {
-            return;
-        }
-
+    boolean notifyPendingReceivedCallback(final Message<T> message, Exception exception) {
         // fetch receivedCallback from queue
         final CompletableFuture<Message<T>> receivedFuture = nextPendingReceive();
         if (receivedFuture == null) {
-            return;
+            if (exception != null) {
+                log.warn("Received exception but no future to callback.", exception);
+                return true;
+            }
+            return getCurrentReceiverQueueSize() == 0;
```

Comment:
From the semantic of method, this condition should not be handle in this method.
And for `ZeroQueueConsumerImpl` only one message is allowed into incoming queueu.

Reply:
got it.